provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -402,24 +402,48 @@ func (c *Client) selectDatastore(
 	if err := c.client.Retrieve(ctx, refs, nil, &datastores); err != nil {
 		return nil, errors.Annotate(err, "retrieving datastore details")
 	}
+
+	if len(datastores) == 0 {
+		c.logger.Errorf("no datastores available")
+		return nil, errors.New("no datastores available")
+	}
+
 	if args.Constraints.RootDiskSource != nil {
 		dsName := *args.Constraints.RootDiskSource
 		c.logger.Debugf("desired datasource %q", dsName)
 		for _, ds := range datastores {
-			c.logger.Debugf("seen %q", ds.Name)
+			c.logger.Debugf("seen %s", ds.Name)
 			if ds.Name == dsName {
+				c.logger.Infof("selecting datastore %s", ds.Name)
 				return &ds, nil
 			}
 		}
 		return nil, errors.Errorf("could not find datastore %q", dsName)
 	}
+
+	var accessibleDatastores []mo.Datastore
 	for _, ds := range datastores {
 		if ds.Summary.Accessible {
-			c.logger.Debugf("using datastore %q", ds.Name)
-			return &ds, nil
+			accessibleDatastores = append(accessibleDatastores, ds)
+		} else {
+			c.logger.Debugf("datastore %s is inaccessible", ds.Name)
 		}
 	}
-	return nil, errors.New("could not find an accessible datastore")
+
+	if len(accessibleDatastores) == 1 {
+		ds := accessibleDatastores[0]
+		c.logger.Infof("selecting datastore %s", ds.Name)
+		return &ds, nil
+	} else if len(accessibleDatastores) > 1 {
+		var datastoreNames []string
+		for _, ds :=range accessibleDatastores {
```

Comment:
I'm probably paranoid, but this might be unpleasant if there are hundreds of datastores. Maybe cap it or don't collect the names if there are more than 10?

Reply:
Good point. I guess the point is that we the user to pick one. We could even just provide a number of datastores, rather than the names.

Reply:
Oh, I definitely think it's more user-friendly to tell the user about the datastores we can see rather than just making them go and work it out themselves. I'm just worried about the extreme cases.